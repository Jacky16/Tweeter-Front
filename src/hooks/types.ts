import { Tweet } from "../types";

export interface TweetsResponse {
  totalPages: number;
  currentPage: number;
  tweets: Tweet[];
}
