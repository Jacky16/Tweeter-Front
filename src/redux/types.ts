import { Tweet, User } from "../types";

export interface UserState extends User {
  isLogged: boolean;
  token: string;
}
export interface UiState {
  isLoading: boolean;
  alert: AlertActionPayloadAction;
  pagination: Pagination;
}

export interface Pagination {
  currentPage: number;
  totalPages: number;
}
export interface TweetState {
  tweets: Tweet[];
  tweet: Tweet;
}

export interface AlertActionPayloadAction {
  severity: "error" | "warning" | "info" | "success";
  message: string;
  isOpen: boolean;
}
