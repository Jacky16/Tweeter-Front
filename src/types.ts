export interface BaseUser {
  username: string;
}

export interface UserRegisterData {
  alias: string;
  password: string;
  email: string;
  username: string;
}

export interface UserLoginData {
  email: string;
  password: string;
}

export interface User extends BaseUser {
  id: string;
  alias: string;
  username: string;
}
export interface JwtPayloadCustom {
  id: string;
  username: string;
  alias: string;
  token: string;
}

export enum TweetCategory {
  comedy,
  science,
  political,
  sports,
  entertainment,
  unknown,
}

export interface TweetStructure {
  id: string;
  description: string;
  category: string;
  dateOfCreation: Date | string;
}
export interface TweetApi extends TweetStructure {
  author: {
    alias: string;
    username: string;
  };
  image: string;
}

export interface TweetData extends TweetStructure {
  image: File;
  visibilityOpen: boolean;
  author: string;
}
export interface Tweet extends TweetStructure {
  username: string;
  alias: string;
  image: string;
}
