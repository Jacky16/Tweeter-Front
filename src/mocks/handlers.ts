// src/mocks/handlers.js
import { rest } from "msw";
import requestsUrl from "../config/requestsUrl";

export const handlers = [
  rest.post(requestsUrl.registerUser, (req, res, ctx) => {
    return res(ctx.status(201), ctx.json({}));
  }),
];
