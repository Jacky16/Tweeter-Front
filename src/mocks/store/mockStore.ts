import { configureStore } from "@reduxjs/toolkit";
import { tweetsReducer } from "../../redux/tweetsSlice/tweetsSlice";
import { TweetState, UiState, UserState } from "../../redux/types";
import { uiReducer } from "../../redux/UiSlice/UiSlice";
import { userReducer } from "../../redux/userSlice/userSlice";

interface MockStoreProps {
  uiPreloadState?: UiState;
  userPreloadState?: UserState;
  tweetsPreloadState?: TweetState;
}
const mockStore = ({
  tweetsPreloadState,
  uiPreloadState,
  userPreloadState,
}: MockStoreProps = {}) =>
  configureStore({
    reducer: {
      ui: uiReducer,
      user: userReducer,
      tweets: tweetsReducer,
    },
    preloadedState: {
      ui: uiPreloadState,
      user: userPreloadState,
      tweets: tweetsPreloadState,
    },
  });

export default mockStore;
