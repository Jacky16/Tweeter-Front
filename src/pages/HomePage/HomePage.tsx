import Grid from "@mui/material/Grid/Grid";
import { useEffect } from "react";
import LoadMoreButton from "../../components/LoadMoreButton/LoadMoreButton";
import TweetCardList from "../../components/TweetCardList/TweetCardList";
import useTweets from "../../hooks/useTweets/useTweets";
import { useAppSelector } from "../../redux/hooks";

const HomePage = () => {
  const {
    pagination: { currentPage, totalPages },
    isLoading,
  } = useAppSelector((state) => state.ui);

  const token = localStorage.getItem("token") as string;

  const tweets = useAppSelector((state) => state.tweets.tweets);

  const { getTweets, loadTweets } = useTweets();

  const isLastPage = currentPage === totalPages;

  useEffect(() => {
    if (currentPage === 1) {
      loadTweets(token);
    } else {
      getTweets(token, currentPage);
    }
  }, [currentPage, getTweets, loadTweets, token]);

  return (
    <Grid
      container
      direction="column"
      justifyContent="center"
      alignItems="center"
      spacing={4}
      paddingY={4}
    >
      <Grid item xs={12}>
        <TweetCardList tweets={tweets} />
      </Grid>
      {!isLastPage && (
        <Grid item>
          <LoadMoreButton
            isLoading={isLoading}
            pagination={{ currentPage, totalPages }}
          />
        </Grid>
      )}
    </Grid>
  );
};

export default HomePage;
