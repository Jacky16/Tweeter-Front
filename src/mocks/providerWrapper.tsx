import { ThemeProvider } from "@mui/system";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "../redux/store";
import mainTheme from "../styles/mainTheme";

interface ProviderWrapperProps {
  children: JSX.Element | JSX.Element[];
}
const ProviderWrapper = ({ children }: ProviderWrapperProps) => {
  return (
    <BrowserRouter>
      <ThemeProvider theme={mainTheme}>
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    </BrowserRouter>
  );
};

export default ProviderWrapper;
