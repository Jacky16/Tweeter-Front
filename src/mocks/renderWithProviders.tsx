import { configureStore, PreloadedState } from "@reduxjs/toolkit";
import { render, RenderOptions } from "@testing-library/react";
import { PropsWithChildren } from "react";
import { Provider } from "react-redux";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { RootState, store } from "../redux/store";
import { tweetsReducer } from "../redux/tweetsSlice/tweetsSlice";
import { uiReducer } from "../redux/UiSlice/UiSlice";
import { userReducer } from "../redux/userSlice/userSlice";
import { InitialEntry } from "@remix-run/router";
interface ExtendedPropsWithChildren extends PropsWithChildren {
  initialEntries?: InitialEntry[];
}
interface ExtendedRenderOptions
  extends Omit<RenderOptions, "queries">,
    ExtendedPropsWithChildren {
  preloadedState?: PreloadedState<RootState>;
  store?: typeof store;
}

const Router = ({
  children,
  initialEntries,
}: ExtendedPropsWithChildren): JSX.Element => {
  return initialEntries ? (
    <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
  ) : (
    <BrowserRouter>{children}</BrowserRouter>
  );
};

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialEntries,
    preloadedState,
    store = configureStore({
      reducer: { ui: uiReducer, user: userReducer, tweets: tweetsReducer },
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <Router initialEntries={initialEntries}>
        <Provider store={store}>{children}</Provider>
      </Router>
    );
  };
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
