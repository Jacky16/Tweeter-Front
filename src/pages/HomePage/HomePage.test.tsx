import { renderWithProviders } from "../../mocks/renderWithProviders";
import mockStore from "../../mocks/store/mockStore";
import { UiState } from "../../redux/types";
import HomePage from "./HomePage";

const mockGetTweets = jest.fn();
const mockLoadTweets = jest.fn();

jest.mock("../../hooks/useTweets/useTweets", () => {
  return () => ({
    getTweets: mockGetTweets,
    loadTweets: mockLoadTweets,
  });
});

describe("Given the HomePage component", () => {
  describe("When it is rendered", () => {
    test("Then should match the snapshot", () => {
      const { container } = renderWithProviders(<HomePage />);
      expect(container).toMatchSnapshot();
    });

    describe("And the current page is 1", () => {
      test("Then loadTweets should be called", () => {
        renderWithProviders(<HomePage />);

        expect(mockLoadTweets).toBeCalled();
      });
    });
    describe("When the currenPage is 2", () => {
      test("Then should getTweets of useTweets should be called", () => {
        const mockUiState: Partial<UiState> = {
          pagination: {
            currentPage: 2,
            totalPages: 3,
          },
        };
        const store = mockStore({ uiPreloadState: mockUiState as UiState });

        renderWithProviders(<HomePage />, { store });

        expect(mockGetTweets).toHaveBeenCalled();
      });
    });
  });
});
