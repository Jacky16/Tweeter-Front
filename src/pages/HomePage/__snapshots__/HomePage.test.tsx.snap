// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the HomePage component When it is rendered Then should match the snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-direction-xs-column css-1jztxya-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <section
        class="css-678bp8-MuiStack-root"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge css-1jd1bad-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        id=":r0:"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTypography-root MuiTypography-button css-lyibsg-MuiTypography-root"
        >
          Load more
        </span>
        <span
          class="MuiButton-endIcon MuiButton-iconSizeLarge css-z4rqyd-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
